baselines/fedpara/README.md

```diff
@@ -0,0 +1,165 @@
+
+```bash
+# Set Python 3.10
+pyenv local 3.10.6
+# Tell poetry to use python 3.10
+poetry env use 3.10.6
+
+# Install the base Poetry environment
+poetry install
+
+# Activate the environment
+poetry shell
+```
+
+## Running the Experiments
+
+Running `FedPara` is easy. You can run it with default parameters directly or by tweaking them directly on the command line. Some command examples are shown below.
+
+```bash  
+# To run fedpara with default parameters
+python -m fedpara.main
+
+# Run for more rounds and a different number of local epochs
+python -m fedpara.main num_rounds=2024 num_epochs=1
+
+# Choose parameterization scheme: lowrank or original (normal weights)
+python -m fedpara.main model.conv_type=standard # or lowrank (default)
+
+# Choosing between non IID and IID scheme
+python -m fedpara.main dataset_config.partition=iid # or non-iid (default)
+
+# Choosing the ratio (lambda) of number of parameters to communicate
+python -m fedpara.main model.ratio=0.1
+
+# Choosing the CIFAR-100 config
+python -m fedpara.main --config-name cifar100 # change settings as shown above if desired
+```
+
+## Expected Results
+
+To reproduce the curves shown below (which correspond to those in Figure 3 in the paper), run the following commands. Experiments running with `model.conv_type=lowrank` correspond to those with `-FedPara` in the legend of the figures below. Those with `model.conv_type=standard` are labelled with the `-orig` (as original) tag.
+
+```bash
+# To run fedpara for non-iid CIFAR-10 on vgg16 for lowrank and original schemes
+python -m fedpara.main --multirun model.conv_type=standard,lowrank 
+# To run fedpara for non-iid CIFAR-100 on vgg16 for lowrank and original schemes
+python -m fedpara.main --config-name cifar100 --multirun model.conv_type=standard,lowrank 
+# To run fedpara for iid CIFAR-10 on vgg16 for lowrank and original schemes
+python -m fedpara.main --multirun model.conv_type=standard,lowrank num_epochs=10 dataset_config.partition=iid 
+# To run fedpara for iid CIFAR-100 on vgg16 for lowrank and original schemes
+python -m fedpara.main --config-name cifar100 --multirun model.conv_type=standard,lowrank num_epochs=10 dataset_config.partition=iid
+```
+
+#### Communication Cost: 
+Communication costs as measured as described in the paper: 
+*"FL evaluation typically measures the required rounds to achieve the target accuracy as communication costs, but we instead assess total transferred bit sizes, 2 ×
+(#participants)×(model size)×(#rounds)"*
+
+
+### CIFAR-100 (Accuracy vs Communication Cost)
+
+| IID | Non-IID |
+|:----:|:----:|
+|![Cifar100 iid](_static/Cifar100_iid.jpeg) | ![Cifar100 non-iid](_static/Cifar100_noniid.jpeg) |
+
+
+### CIFAR-10 (Accuracy vs Communication Cost)
+
+| IID | Non-IID |
+|:----:|:----:|
+|![CIFAR10 iid](_static/Cifar10_iid.jpeg) | ![CIFAR10 non-iid](_static/Cifar10_noniid.jpeg) |
```

Comment:
At the end of the readme you can add a new section showing how the personalisation experiments with MNIST can be executed. Even if they are not fully working, this would give others a good starting point. But please add a comment right under the title of the section stating that it has been quite hard reproducing the results (maybe give some details of what you tried, what would be the next steps to investigate etc)